docs/ref-manual/cvl/invariants.md

```diff
@@ -28,29 +35,269 @@ preserved_block ::= "preserved"
+method will be skipped.
+
+```{note}
+If there is a {ref}`preserved block <preserved>` for a method it will be
+verified even if it should be filtered out.
+```
+
+(preserved)=
 Preserved blocks
 ----------------
 
-```{todo}
-This feature is currently undocumented.
+Often, the preservation of an invariant depends on another invariant, or on an
+external assumption about the system.  These assumptions can be written in
+`preserved` blocks.
+
+```{caution}
+Adding `require` statements to preserved blocks can be a source of
+{term}`unsound`ness, since the invariants are only guaranteed to hold if the
+requirements are true for every method invocation.
+```
+
+Recall that the Prover checks that a method preserves an invariant by first
+requiring the invariant (the pre-state check), then executing the method, and
+then asserting the invariant (the post-state check).  Preserved blocks are
+executed after the pre-state check but before executing the method.  They
+usually consist of `require` or `requireInvariant` statements, although other
+commands are also possible.
+
+Preserved blocks are listed after the invariant expression (and after the filter
+block, if any), inside a set of curly braces (`{ ... }`).  Each preserved block
+consists of the keyword `preserved` followed by an optional method signature,
+an optional `with` declaration, and finally the block of commands to execute.
+
+If a preserved block specifies a method signature, the signature should
+match one of the contract methods, and the preserved block only applies when
+checking preservation of that contract method.  The arguments of the method are
+in scope within the preserved block.  If there is no method signature, the
+preserved is a default block that applies to all methods that don't have a
+specific preserved block.
+
+The `with` declaration is used to give a name to the {term}`environment` used
+while invoking the method.  It can be used to restrict the transactions that are
+considered.  For example, the following preserved block rules out
+counterexamples where the `msg.sender` is the 0 address:
+
+```cvl
+invariant zero_address_has_no_balance()
+    balanceOf(0) == 0
+    { preserved with (env e) { require e.msg.sender != 0; } }
+```
+
+The variables defined as parameters to the invariant are also available in
+preserved blocks, which allows restricting the arguments that are considered
+when checking that a method preserves an invariant.
+
+A common source of confusion is the difference between `env` parameters
+to an invariant and the `env` variables defined by the `with` declaration.
+Compare the following to the previous example:
+
+```cvl
+invariant zero_address_has_no_balance_v2(env e)
+    balanceOf(e, 0) == 0
+    { preserved { require e.msg.sender != 0; } }
+```
+
+In this example, we require the `msg.sender` argument to `balanceOf` to be
+nonzero, but makes no restrictions on the environment for the call to the method
+we are checking for preservation.
```

Comment:
This requires some more explanation. I would show an example code snippet where this confusion leads to the wrong result

Reply:
I expanded the example as suggested.